### PR TITLE
[JSON-API] Remove dependency on user provided tokens for the perf runner & refactor main

### PR DESCRIPTION
**Pull request by realvictorprm**
Fix for https://github.com/digital-asset/daml/issues/13105

changelog_begin
changelog_end

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala [hunk 1]
```diff
@@ -116,33 +118,62 @@ object Main extends StrictLogging {
...
+    // after we started the ledger.
+    val (userIdOpt, ledgerId) =
+      HttpService
+        .decodeJwt(config.jwt)
+        .leftMap(_.message)
+        .flatMap { decodedJwt =>
+          CreateFromUserToken
+            .parseAndDecodeUserToken(decodedJwt)
+            .bimap(
+              error => s"This is neither a custom token nor an user token: ${error.message}",
```
Comment (ray-roestenburg-da) [inline ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala:154]:
```suggestion
              error => s"This is neither a custom token nor a user token: ${error.message}",
```

Comment (realvictorprm) [reply ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala:154]: That made me thinking, thanks!

File: ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala [hunk 2]
```diff
@@ -116,33 +118,62 @@ object Main extends StrictLogging {
...
+        main2(config)
       ): ET[ExitCode]
 
     } yield exitCode
 
     et.run
   }
 
-  private def main2(ledgerId: LedgerId, config: Config[String])(implicit
+  private def main2(config: Config[String])(implicit
```
Comment (ray-roestenburg-da) [inline ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala:135]: main1 and main2 are a bit strange names (for a newcomer like me) can this be more descriptive?

Comment (realvictorprm) [reply ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala:135]: I agree... Had the exact same thought already too

File: ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala [hunk 3]
```diff
@@ -116,33 +118,62 @@ object Main extends StrictLogging {
...
+                    ),
+                  )
+              },
+            )
+        }
+        .fold((error: String) => throw new Exception(error), identity)
+    withLedger(config.dars, ledgerId.unwrap) { (ledgerPort, ledgerClient, _) =>
+      // For a user token to work the user has to be created beforehand.
+      userIdOpt.foreach { userId =>
+        ledgerClient.userManagementClient.createUser(
```
Comment (ray-roestenburg-da) [inline ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala:172]: I think you might be ignoring result of future here.

File: ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala [hunk 4]
```diff
@@ -116,33 +118,62 @@ object Main extends StrictLogging {
...
+            .bimap(
+              error => s"This is neither a custom token nor a user token: ${error.message}",
+              {
+                case StandardJWTPayload(userId, _, _) =>
+                  (Some(userId), LedgerId("perf-runner"))
+                case CustomDamlJWTPayload(ledgerId, _, _, _, _, _, _) =>
+                  (
+                    None,
+                    LedgerId(
+                      ledgerId.getOrElse(throw new Exception("No ledger id given in custom token"))
```
Comment (ray-roestenburg-da) [inline ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala:162]: Rather not throw exceptions, but lift disjunction into future, and have a simple for comp on everything, failing fast.

File: ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala [hunk 5]
```diff
@@ -116,51 +120,81 @@ object Main extends StrictLogging {
...
+        .flatMap { decodedJwt =>
+          CreateFromUserToken
+            .parseAndDecodeUserToken(decodedJwt)
+            .bimap(
+              error => s"This is neither a custom token nor a user token: ${error.message}",
+              {
+                case StandardJWTPayload(userId, _, _) =>
+                  (Some(userId), LedgerId("perf-runner"))
+                case CustomDamlJWTPayload(ledgerIdPayload, _, _, _, _, _, _) =>
+                  (None, LedgerId(ledgerIdPayload.getOrElse("perf-runner")))
```
Comment (stefanobaghino-da) [inline ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala:161]:
```suggestion
                case payload: StandardJWTPayload =>
                  (Some(payload.userId), LedgerId("perf-runner"))
                case payload: CustomDamlJWTPayload =>
                  (None, LedgerId(payload.ledgerId.getOrElse("perf-runner")))
```
Minor: the main advantage of using this approach is that if the `case class` adds or removes fields there won't be any change necessary here.

File: ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala [hunk 6]
```diff
@@ -116,51 +120,81 @@ object Main extends StrictLogging {
...
+      (userIdOpt, ledgerId) = userAndLedgerId
+      exitCode <- withLedger(config.dars, ledgerId.unwrap) { (ledgerPort, ledgerClient, _) =>
+        // For a user token to work the user has to be created beforehand.
+        for {
+          _ <-
+            userIdOpt.traverse { userId =>
+              ledgerClient.userManagementClient
+                .createUser(
+                  User(UserId.assertFromString(userId), None),
+                  List(CanActAs(Party.assertFromString("Alice"))),
```
Comment (stefanobaghino-da) [inline ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala:175]: Not for this PR (unless it's an easy fix), but we should probably stop relying on hard-coded party names before it _really_ becomes a problem.

File: ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala [hunk 7]
```diff
@@ -116,51 +120,81 @@ object Main extends StrictLogging {
...
+                      logger.info(s"Report directory: ${dir.toAbsolutePath}")
+                      exitCode
+                    }
+                }: Future[ExitCode]
+            }
+          }
+        } yield res
       }
-    }
+    } yield exitCode
```
Comment (stefanobaghino-da) [inline ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala:197]: There's a nested `for`-comprehension and a lot of indentation here. Breaking the steps into methods could improve the overall readability.

File: ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala [hunk 8]
```diff
@@ -323,4 +93,84 @@ object Main extends StrictLogging {
...
+                      }
+                  }
+              }
+            }
+          }
+        })
+        waitForResult(logCompletion(res.run), config.maxDuration.getOrElse(Duration.Inf))
+    }
+    terminate()
+    sys.exit(exitCode.unwrap)
```
Comment (ray-roestenburg-da) [inline ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala:174]: why is this needed?

Comment (ray-roestenburg-da) [reply ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala:174]: (I guess something else running this depends on specific exit codes)

File: ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/OracleRunner.scala [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  def start(): Try[St] = Try {
+    if (useDefaultUser) OracleAround.createOrReuseUser(defaultUser)
+    else OracleAround.createNewUniqueRandomUser()
+  }
+
+  def jdbcConfig(user: St): JdbcConfig = {
+    import DbStartupMode._
+    val startupMode: DbStartupMode = if (retainData) CreateIfNeededAndStart else CreateAndStart
+    JdbcConfig(
+      dbutils.JdbcConfig(
```
Comment (stefanobaghino-da) [inline ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/OracleRunner.scala:29]: The naming seems a bit confusing. You can tell them apart from the package name of course, but still likely to be causing confusion.

File: ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala [hunk 10]
```diff
@@ -323,4 +90,87 @@ object Main extends StrictLogging {
...
+          }
+        }
+      }
+
+    val exitCode: ExitCode = Config.parseConfig(args) match {
+      case None =>
+        // error is printed out by scopt
+        ExitCode.InvalidUsage
+      case Some(config) =>
+        logger.info(s"$config")
```
Comment (stefanobaghino-da) [inline ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala:145]:
```suggestion
        logger.info(config)
```

**Review by stefanobaghino-da — APPROVED** [review]

File: ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/QueryStoreBracket.scala [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+import scala.util.Try
+
+private[this] final case class QueryStoreBracket[S, D](
+    state: () => S,
+    start: S => Try[D],
+    config: (S, D) => JdbcConfig,
+    stop: (S, D) => Try[Unit],
+)
+
+private[this] object QueryStoreBracket {
```
[the remark this answers is not in the document]
Comment (realvictorprm) [reply ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/QueryStoreBracket.scala:23]: honestly, I will just remove it because I'm don't think it is needed. After all I was initially just rearranging code and not touching all parts.

File: ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Config.scala [hunk 12]
```diff
@@ -40,21 +36,11 @@ private[perf] object Config {
     Config[String](
       scenario = "",
       dars = List.empty,
-      jwt = Jwt(""),
       reportsDir = new File(""),
       maxDuration = None,
       queryStoreIndex = QueryStoreIndex.No,
     )
 
-  implicit val configInstance: Traverse[Config] = new Traverse[Config] {
```
Comment (ray-roestenburg-da) [inline ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Config.scala:49]: happy this is removed, I don't see the value in a traverse instance on a `Config`?

File: ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala [hunk 13]
```diff
@@ -49,232 +35,14 @@ object Main extends StrictLogging {
...
-          ConnectionPool.PoolSize.Production,
-        ),
-      startMode = DbStartupMode.CreateOnly,
-    )
-
-  private def resolveSimulationClass(str: String): Throwable \/ Class[_ <: Simulation] = {
-    try {
+  private def resolveSimulationClass(
+      str: String
+  )(implicit ec: ExecutionContext): Future[Class[_ <: Simulation]] = {
```
Comment (ray-roestenburg-da) [inline ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala:40]: Of course this existed already, is instantiation by reflection really necessary? Rather avoid doing that, if necessary pattern match on some config and create the actual class that is needed, which would be far more readable.

File: ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala [hunk 14]
```diff
@@ -323,4 +90,87 @@ object Main extends StrictLogging {
...
+    val terminationTimeout: FiniteDuration = 30.seconds
+
+    implicit val asys: ActorSystem = ActorSystem(name)
+    implicit val mat: Materializer = Materializer(asys)
+    implicit val aesf: ExecutionSequencerFactory =
+      new AkkaExecutionSequencerPool(poolName = name, terminationTimeout = terminationTimeout)
+    implicit val elg: EventLoopGroup = Transports.newEventLoopGroup(true, 0, "gatling")
+    implicit val ec: ExecutionContext = asys.dispatcher
+
+    def terminate(): Unit = {
```
Comment (ray-roestenburg-da) [inline ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala:105]: I would prefer not to use discard here, everything is side effecting, it's obvious. If fatal warnings can;t be removed maybe just suppress the warning inside terminate.

File: ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala [hunk 15]
```diff
@@ -323,4 +90,87 @@ object Main extends StrictLogging {
...
+      discard {
+        future.addListener((f: io.netty.util.concurrent.Future[_]) =>
+          discard { promise.complete(Try(f.get).map(_ => ())) }
+        )
+      }
+      discard { Await.result(promise.future, terminationTimeout) }
+    }
+
+    def runScenario(config: Config[String]) =
+      resolveSimulationClass(config.scenario).flatMap { _ =>
```
Comment (ray-roestenburg-da) [inline ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala:118]: the instance created by reflection is discarded, how does this work?

File: ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/QueryStoreBracket.scala [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+import com.daml.dbutils.ConnectionPool
+import com.daml.http.dbbackend.{DbStartupMode, JdbcConfig}
+import com.daml.http.perf.Config.QueryStoreIndex
+import com.daml.http.util.FutureUtil.toFuture
+import com.daml.testing.postgresql.PostgresDatabase
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Try
+
+private[this] final case class QueryStoreBracket[S, D](
```
Comment (ray-roestenburg-da) [inline ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/QueryStoreBracket.scala:16]: I think this can be far more easily done with a trait and implementations.

File: ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/QueryStoreBracket.scala [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+    case QueryStoreIndex.Oracle => Some(Oracle)
+  }
+
+  def withJsonApiJdbcConfig[A](
+      jsonApiQueryStoreEnabled: QueryStoreIndex
+  )(
+      fn: Option[JdbcConfig] => Future[A]
+  )(implicit
+      ec: ExecutionContext
+  ): Future[A] = QueryStoreBracket lookup jsonApiQueryStoreEnabled match {
```
Comment (ray-roestenburg-da) [inline ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/QueryStoreBracket.scala:65]: Almost looks like we are emulating virtual dispatch here. Rather just encode the steps in the trait, and have an Oracle and Postgress implemenation fill in details.

File: ci/cron/daily-compat.yml [hunk 18]
```diff
@@ -161,8 +159,7 @@ jobs:
             bazel run //ledger-service/http-json-perf:http-json-perf-binary-ce -- \
             --scenario=${scenario} \
             --dars=${DAR} \
-            --reports-dir=${OUT} \
-            --jwt=${JWT}
+            --reports-dir=${OUT}
```
Comment (realvictorprm) [inline ci/cron/daily-compat.yml:162]: Tested locally with
```
set -euo pipefail
eval "$(dev-env/bin/dade assist)"
SCENARIOS="\
com.daml.http.perf.scenario.CreateCommand \
com.daml.http.perf.scenario.ExerciseCommand \
com.daml.http.perf.scenario.CreateAndExerciseCommand \
com.daml.http.perf.scenario.AsyncQueryConstantAcs \
com.daml.http.perf.scenario.SyncQueryConstantAcs \
com.daml.http.perf.scenario.SyncQueryNewAcs \
com.daml.http.perf.scenario.SyncQueryVariableAcs \
"
bazel build //docs:quickstart-model
DAR="${PWD}/bazel-bin/docs/quickstart-model.dar"

for scenario in $SCENARIOS; do
  bazel run //ledger-service/http-json-perf:http-json-perf-binary-ce -- \
  --scenario=${scenario} \
  --dars=${DAR}
done
```

to see if the script is running successfully and the output I got looked fine so far (didn't run it completely because the perf tests take that long).

File: ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/QueryStoreBracket.scala (hunk 11 above)
[the remark this answers is not in the document]
Comment (realvictorprm) [reply ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/QueryStoreBracket.scala:23]: will do all that in follow up pr's

**Comment by realvictorprm** [comment]
Will enable auto merge now as I think all should be fine now :)
